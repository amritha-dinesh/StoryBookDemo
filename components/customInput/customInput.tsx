import React, {useState} from 'react';
import {
  View,
  TextInput,
  StyleSheet,
  SafeAreaView,
  Platform,
  Animated,
  Image,
  StyleProp,
  TextStyle,
  TouchableOpacity,
} from 'react-native';
import {black, white, blue500} from '../styles/themes/colors';

const BORDER_RADIUS = 5;

interface CustomInputProps {
  mode?: 'outlined' | 'rounded' | 'standard';
  disabled?: boolean;
  passwordForm?: boolean;
  keyboardType?:
    | 'default'
    | 'email-address'
    | 'numeric'
    | 'phone-pad'
    | 'number-pad'
    | 'decimal-pad'
    | 'ascii-capable'
    | 'numbers-and-punctuation'
    | 'url'
    | 'name-phone-pad'
    | 'twitter'
    | 'web-search';
  label?: string;
  placeholder?: string;
  num?: boolean;
  testID?: string;
  onChangeText?: (text: string) => void;
  value?: string;
  rightIcon?: string | React.ReactElement;
  leftIcon?: React.ReactElement;
  placeholderTextColor?: string;
  numberOfLines?: number;
  autoCapitalize?: 'none' | 'sentences' | 'words' | 'characters' | undefined;
  multiline?: boolean;
  textAlign?: 'left' | 'center' | 'right';
  onFocus?: () => void;
  onBlur?: () => void;
  autoFocus?: boolean;
  defaultValue?: string | undefined;
  selectionColor?: string;
  labelColor?: string;
  borderRadius?: number;
  style?: StyleProp<TextStyle>;
}

export const CustomInput: React.FC<CustomInputProps> = ({
  mode = 'filled',
  style,
  disabled = false,
  passwordForm = false,
  keyboardType = 'default',
  label = 'Label Name',
  placeholder = 'placeholder',
  testID = 'custom-input',
  onChangeText,
  value,
  rightIcon,
  leftIcon,
  placeholderTextColor = 'gray',
  numberOfLines,
  autoCapitalize,
  multiline,
  textAlign,
  onFocus,
  onBlur,
  autoFocus = false,
  defaultValue,
  selectionColor,
  labelColor = black,
  borderRadius = BORDER_RADIUS,
  ...props
}) => {
  const animatedLabelStyle = new Animated.Value(1);
  const [focus, setFocus] = useState(false);
  const [icon, setIcon] = useState(false);
  return (
    <SafeAreaView>
      <View style={styles.container}>
        <View style={styles.subContainer}>
          <Animated.Text
            {...props}
            testID={'label'}
            style={[
              styles.labelStyle,
              animatedLabelStyle,
              {color: labelColor},
            ]}>
            {label}
          </Animated.Text>
          <View style={styles.inputContainer}>
            {leftIcon && <TouchableOpacity>{leftIcon}</TouchableOpacity>}
            <TextInput
              {...props}
              testID={testID}
              defaultValue={defaultValue}
              onPressIn={onBlur}
              onPressOut={onFocus}
              autoCapitalize={autoCapitalize}
              numberOfLines={numberOfLines}
              placeholderTextColor={placeholderTextColor}
              autoFocus={autoFocus}
              selectionColor={selectionColor}
              onFocus={() => {
                if (disabled === true) {
                  setFocus(false);
                } else {
                  setFocus(true);
                }
              }}
              onBlur={() => {
                setFocus(false);
              }}
              multiline={multiline}
              keyboardType={keyboardType}
              secureTextEntry={passwordForm && !icon ? true : false}
              textAlign={textAlign}
              editable={disabled ? false : true}
              value={value}
              onChangeText={onChangeText}
              placeholder={placeholder}
              style={[
                styles.input,
                leftIcon && styles.leftIconInput,
                mode === 'outlined'
                  ? [styles.outlined, {borderRadius: borderRadius}]
                  : mode === 'rounded'
                  ? styles.rounded
                  : styles.standards,
                focus === true && styles.focused,
                style,
              ]}
            />
            {rightIcon && (
              <TouchableOpacity onPress={() => setIcon(!icon)}>
                {rightIcon}
              </TouchableOpacity>
            )}
            {passwordForm && (
              <TouchableOpacity
                onPress={() => {
                  setIcon(!icon);
                }}>
                <Image
                  source={
                    !icon
                      ? require('../assets/eye.png')
                      : require('../assets/eye-off.png')
                  }
                  style={styles.icon}
                />
              </TouchableOpacity>
            )}
          </View>
        </View>
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    backgroundColor: white,
    width: '100%',
  },
  outlined: {
    borderWidth: Platform.OS === 'ios' ? 1 : 0.5,
    backgroundColor: 'transparent',
  },
  rounded: {
    borderWidth: Platform.OS === 'ios' ? 1 : 0.5,
    backgroundColor: 'transparent',
    borderRadius: 22,
  },
  focused: {
    borderColor: blue500,
  },
  standards: {
    borderBottomWidth: Platform.OS === 'ios' ? 1 : 0.5,
    backgroundColor: 'transparent',
  },
  leftIconInput: {
    paddingLeft: 40,
  },
  input: {
    width: '100%',
    height: 44,
    padding: 10,
    marginTop: 20,
    marginBottom: 10,
  },
  inputContainer: {
    flexDirection: 'row',
  },
  icon: {
    height: 20,
    width: 20,
    marginTop: 33,
    right: 30,
  },
  leftIcon: {
    height: 20,
    width: 20,
    marginTop: 33,
    position: 'absolute',
    left: 10,
  },
  labelStyle: {
    position: 'absolute',
  },
  subContainer: {
    alignSelf: 'center',
    width: '90%',
  },
  main: {
    flex: 1,
  },
});
export default CustomInput;
