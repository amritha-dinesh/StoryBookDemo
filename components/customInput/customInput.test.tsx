/**
 * @format
 */

import { Platform } from "react-native";
import React from "react";
import renderer from "react-test-renderer";
import { render } from "@testing-library/react-native";
import CustomInput from "./customInput";
// Note: test renderer must be required after react-native.

describe("CustomInput Component", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<CustomInput />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly find", () => {
    const { getByTestId } = render(<CustomInput />);
    const customInputComponent = getByTestId("custom-input");
    expect(customInputComponent).toBeDefined();
  });

  it("renders custom input mode by default", () => {
    const { getByTestId } = render(<CustomInput />);
    const filledStyle = {
      borderBottomWidth: Platform.OS === "ios" ? 1 : 0.5,
      backgroundColor: "#e8e8e8",
    };
    const customInputComponent = getByTestId("custom-input");
    expect(customInputComponent.props.style[1]).toEqual(filledStyle);
  });

  it("renders disabled cusom input when mode prop is outlined", () => {
    const { getByTestId } = render(<CustomInput mode="outlined" />);
    const outlinedStyle = {
      borderWidth: Platform.OS === "ios" ? 1 : 0.5,
      backgroundColor: "transparent",
    };
    const customInputComponent = getByTestId("custom-input");
    expect(customInputComponent.props.style[1]).toEqual(outlinedStyle);
  });

  it("enders disabled cusom input when mode prop is standard", () => {
    const { getByTestId } = render(<CustomInput mode="standard" />);
    const standardStyle = {
      borderBottomWidth: Platform.OS === "ios" ? 1 : 0.5,
      backgroundColor: "transparent",
    };
    const customInputComponent = getByTestId("custom-input");
    expect(customInputComponent.props.style[1]).toEqual(standardStyle);
  });

  it("renders enabled custom input by default", () => {
    const { getByTestId } = render(<CustomInput />);
    const customInputComponent = getByTestId("custom-input");
    expect(customInputComponent.props.editable).toBeTruthy();
  });

  it("renders disabled cusom input when disabled prop is true", () => {
    const { getByTestId } = render(<CustomInput disabled={true} />);
    const customInputComponent = getByTestId("custom-input");
    expect(customInputComponent.props.editable).toBeFalsy();
  });

  it("renders custom input passwordForm by default", () => {
    const { getByTestId } = render(<CustomInput />);
    const customInputComponent = getByTestId("custom-input");
    expect(customInputComponent.props.secureTextEntry).toBeFalsy();
  });

  it("renders disabled cusom input when passwordForm prop is true", () => {
    const { getByTestId } = render(<CustomInput passwordForm={true} />);
    const customInputComponent = getByTestId("custom-input");
    expect(customInputComponent.props.secureTextEntry).toBeTruthy();
  });

  it("renders enabled custom input keyboardType by default", () => {
    const { getByTestId } = render(<CustomInput />);
    const keyboardType = "default";
    const customInputComponent = getByTestId("custom-input");
    expect(customInputComponent.props.keyboardType).toBe(keyboardType);
  });

  it("renders disabled cusom input when keyboardType prop is numeric", () => {
    const { getByTestId } = render(<CustomInput keyboardType="numeric" />);
    const keyboardType = "numeric";
    const customInputComponent = getByTestId("custom-input");
    expect(customInputComponent.props.keyboardType).toBe(keyboardType);
  });

  it("renders disabled cusom input label by default", () => {
    const { getByTestId } = render(<CustomInput />);
    const label = "Label Name";
    const customInputComponent = getByTestId("label");
    expect(customInputComponent.props.children).toBe(label);
  });

  it("renders disabled cusom input placeholder by default", () => {
    const { getByTestId } = render(<CustomInput />);
    const placeholder = "placeholder";
    const customInputComponent = getByTestId("custom-input");
    expect(customInputComponent.props.placeholder).toBe(placeholder);
  });
});
