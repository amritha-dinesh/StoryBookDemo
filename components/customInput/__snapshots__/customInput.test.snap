// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomInput Component renders correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#ffffff",
        "flex": 1,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "alignSelf": "center",
          "width": "90%",
        }
      }
    >
      <Text
        collapsable={false}
        style={
          {
            "_animation": null,
            "_children": [],
            "_listeners": {},
            "_offset": 0,
            "_startingValue": 1,
            "_value": 1,
            "color": "black",
            "position": "absolute",
          }
        }
        testID="label"
      >
        Label Name
      </Text>
      <View
        style={
          {
            "flexDirection": "row",
          }
        }
      >
        <TextInput
          autoFocus={false}
          editable={true}
          keyboardType="default"
          onBlur={[Function]}
          onFocus={[Function]}
          placeholder="placeholder"
          placeholderTextColor="gray"
          secureTextEntry={false}
          style={
            [
              {
                "height": 44,
                "marginBottom": 10,
                "marginTop": 20,
                "padding": 10,
                "width": "100%",
              },
              {
                "backgroundColor": "#e8e8e8",
                "borderBottomWidth": 1,
              },
              false,
            ]
          }
          testID="custom-input"
        />
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
